eppo_core/src/bandits/eval.rs

```diff
@@ -0,0 +1,421 @@
+use std::collections::HashMap;
+
+use chrono::Utc;
+use serde::Deserialize;
+use serde::Serialize;
+
+use crate::sharder::Md5Sharder;
+use crate::sharder::Sharder;
+use crate::ufc::Assignment;
+use crate::ufc::AssignmentEvent;
+use crate::ufc::AssignmentValue;
+use crate::ufc::VariationType;
+use crate::Configuration;
+use crate::ContextAttributes;
+
+use super::event::BanditEvent;
+use super::BanditCategoricalAttributeCoefficient;
+use super::BanditModelData;
+use super::BanditNumericAttributeCoefficient;
+
+#[derive(Debug)]
+struct BanditEvaluationDetails {
+    pub flag_key: String,
+    pub subject_key: String,
+    pub subject_attributes: ContextAttributes,
+    /// Selected action.
+    pub action_key: String,
+    /// Attributes of the selected action.
+    pub action_attributes: ContextAttributes,
+    /// Score of the selected action.
+    pub action_score: f64,
+    pub action_weight: f64,
+    pub gamma: f64,
+    /// Distance between best and selected actions' scores.
+    pub optimality_gap: f64,
+}
+
+struct Action<'a> {
+    key: &'a str,
+    attributes: &'a ContextAttributes,
+}
+
+/// Result of evaluating a bandit.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct BanditResult {
+    /// Selected variation from the feature flag.
+    variation: String,
+    /// Selected action if any.
+    action: Option<String>,
+    /// Flag assignment event that needs to be logged to analytics storage.
+    assignment_event: Option<AssignmentEvent>,
+    /// Bandit assignment event that needs to be logged to analytics storage.
+    bandit_event: Option<BanditEvent>,
+}
+
+impl Configuration {
+    /// Evaluate the specified string feature flag for the given subject. If resulting variation is
+    /// a bandit, evaluate the bandit to return the action.
+    pub fn get_bandit_action(
+        &self,
+        flag_key: &str,
+        subject_key: &str,
+        subject_attributes: &ContextAttributes,
+        actions: &HashMap<String, ContextAttributes>,
+        default_variation: &str,
+    ) -> BanditResult {
+        let assignment = self
+            .get_assignment(
+                flag_key,
+                subject_key,
+                &subject_attributes.to_generic_attributes(),
+                Some(VariationType::String),
+            )
+            .unwrap_or_default()
+            .unwrap_or_else(|| Assignment {
+                value: AssignmentValue::String(default_variation.to_owned()),
+                event: None,
+            });
```

Comment:
One deviation from other SDKs: because of how `eppo_core` is structured, it is possible to evaluate assignments without triggering assignment logging. I therefore removed an early return in case it's a bandit flag but actions are missing and evaluate the assignment immediately.
- In case assignment resolves to non-bandit variation, we return the variation and log the assignment.
- If it's a bandit variation and actions are missing, bandit evaluation will fail and we'll return default variation and skip assignment logging.
- Otherwise, we return variation+action + log the assignment and bandit.

Reply:
I think you could mirror the other SDKs if you stored top-level which flags have which bandits. Then you know if a flag involves bandits or not from the get go and can check if assignments is empty.

I do agree it's kind of a pain though.

Reply:
Yeah, I store that mapping and I can easily mirror the behavior, but I thought that this behavior is more desirable from the user perspective (they receive correct variation instead of the default one).